File: myExpenses/src/main/java/org/totschnig/myexpenses/util/licence/LicenceStatus.java
package org.totschnig.myexpenses.util.licence;

import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.model.ContribFeature;

import java.util.Locale;

import androidx.annotation.Nullable;

public enum LicenceStatus {
  CONTRIB(R.string.contrib_key), EXTENDED(R.string.extended_key), EXTENDED_FALLBACK(R.string.extended_key), PROFESSIONAL(R.string.professional_key) {
    @Override
    public boolean isUpgradeable() {
      return false;
    }
  };

  private final int resId;

  LicenceStatus(int resId) {
    this.resId = resId;
  }

  public int getResId() {
    return resId;
  }

  public boolean greaterOrEqual(@Nullable LicenceStatus other) {
    return other == null || compareTo(other) >= 0;
  }

  public boolean covers(ContribFeature contribFeature) {
    if (contribFeature == null) return true;
    return greaterOrEqual(contribFeature.getLicenceStatus());
  }

  public boolean isUpgradeable() {
    return true;
  }

  /**
   * for historical reasons, skus for Contrib used "premium"
   */
  public String toSkuType() {
    if (this == LicenceStatus.CONTRIB) {
      return "premium";
    }
    return name().toLowerCase(Locale.ROOT);
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/util/licence/LicenceStatus.kt
package org.totschnig.myexpenses.util.licence

import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.model.ContribFeature

enum class LicenceStatus(val resId: Int, val isUpgradeable: Boolean = true) {
    CONTRIB(R.string.contrib_key),
    EXTENDED(R.string.extended_key),
    EXTENDED_FALLBACK(R.string.extended_key),
    PROFESSIONAL(R.string.professional_key, false);

    val color: Int
        @ColorRes get() = when (this) {
            CONTRIB -> R.color.premium_licence
            EXTENDED, EXTENDED_FALLBACK -> R.color.extended_licence
            PROFESSIONAL -> R.color.professional_licence
        }

    fun greaterOrEqual(other: LicenceStatus?): Boolean {
        return other == null || this >= other
    }

    fun covers(contribFeature: ContribFeature?): Boolean {
        return if (contribFeature == null) true else this >= contribFeature.licenceStatus
    }

    /**
     * for historical reasons, skus for Contrib used "premium"
     */
    fun toSkuType(): String {
        return if (this === CONTRIB) {
            "premium"
        } else name.lowercase()
    }
}